node/peers.js

```diff
@@ -144,23 +147,33 @@ TChannelPeers.prototype.clear = function clear() {
         for (var i = 0; i < names.length; i++) {
             var subChannel = self.channel.subChannels[names[i]];
             subChannel.peers._map = Object.create(null);
+            subChannel.peers._keys = [];
         }
     }
     self._map = Object.create(null);
+    self._keys = [];
     return vals;
 };
 
 TChannelPeers.prototype.delete = function del(hostPort) {
     var self = this;
     var peer = self._map[hostPort];
+
+    if (!peer) {
+        return;
+    }
+
     if (self.channel.subChannels) {
         var names = Object.keys(self.channel.subChannels);
         for (var i = 0; i < names.length; i++) {
             var subChannel = self.channel.subChannels[names[i]];
-            delete subChannel.peers._map[hostPort];
+            subChannel.peers.delete(hostPort);
         }
     }
     delete self._map[hostPort];
+    var index = self._keys.indexOf(hostPort);
+    self._keys.splice(index, 1);
```

Comment:
actually, since the order is unimportant, we can:

```js
self._keys[index] = self._keys[self._keys.index.length - 1];
self._keys.length--;
```

Look Ma', no function calls and O(1).